File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/go/pkg/resolver/resolver_test.go
package resolver

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/ThirdKeyAi/schemapin/go/pkg/bundle"
	"github.com/ThirdKeyAi/schemapin/go/pkg/discovery"
	"github.com/ThirdKeyAi/schemapin/go/pkg/revocation"
)

func makeBundle() *bundle.SchemaPinTrustBundle {
	b := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	b.Documents = append(b.Documents, bundle.BundledDiscovery{
		Domain: "example.com",
		WellKnown: discovery.WellKnownResponse{
			SchemaVersion: "1.2",
			DeveloperName: "Test Dev",
			PublicKeyPEM:  "-----BEGIN PUBLIC KEY-----\ntest\n-----END PUBLIC KEY-----",
		},
	})

	rev := revocation.BuildRevocationDocument("example.com")
	revocation.AddRevokedKey(rev, "sha256:old", revocation.ReasonSuperseded)
	b.Revocations = append(b.Revocations, *rev)

	return b
}

func TestTrustBundleResolverDiscoveryHit(t *testing.T) {
	resolver := NewTrustBundleResolver(makeBundle())
	disc, err := resolver.ResolveDiscovery("example.com")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if disc.DeveloperName != "Test Dev" {
		t.Errorf("expected Test Dev, got %s", disc.DeveloperName)
	}
}

func TestTrustBundleResolverDiscoveryMiss(t *testing.T) {
	resolver := NewTrustBundleResolver(makeBundle())
	_, err := resolver.ResolveDiscovery("unknown.com")
	if err == nil {
		t.Error("expected error for unknown domain")
	}
}

func TestTrustBundleResolverRevocation(t *testing.T) {
	resolver := NewTrustBundleResolver(makeBundle())
	disc, _ := resolver.ResolveDiscovery("example.com")
	rev, err := resolver.ResolveRevocation("example.com", disc)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if rev == nil {
		t.Fatal("expected non-nil revocation")
	}
	if rev.Domain != "example.com" {
		t.Errorf("expected domain example.com, got %s", rev.Domain)
	}
}

func TestTrustBundleResolverFromJSON(t *testing.T) {
	b := makeBundle()
	data, err := json.Marshal(b)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	resolver, err := FromJSON(string(data))
	if err != nil {
		t.Fatalf("failed to create from JSON: %v", err)
	}

	disc, err := resolver.ResolveDiscovery("example.com")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if disc.DeveloperName != "Test Dev" {
		t.Errorf("expected Test Dev, got %s", disc.DeveloperName)
	}
}

func TestLocalFileResolverDiscovery(t *testing.T) {
	tmpDir := t.TempDir()
	wellKnown := discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "File Dev",
		PublicKeyPEM:  "PEM_DATA",
	}
	data, _ := json.Marshal(wellKnown)
	if err := os.WriteFile(filepath.Join(tmpDir, "example.com.json"), data, 0644); err != nil {
		t.Fatalf("failed to write file: %v", err)
	}

	resolver := NewLocalFileResolver(tmpDir, "")
	disc, err := resolver.ResolveDiscovery("example.com")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if disc.DeveloperName != "File Dev" {
		t.Errorf("expected File Dev, got %s", disc.DeveloperName)
	}
}

func TestLocalFileResolverDiscoveryMissing(t *testing.T) {
	tmpDir := t.TempDir()
	resolver := NewLocalFileResolver(tmpDir, "")
	_, err := resolver.ResolveDiscovery("missing.com")
	if err == nil {
		t.Error("expected error for missing file")
	}
}

func TestLocalFileResolverRevocation(t *testing.T) {
	tmpDir := t.TempDir()
	rev := revocation.BuildRevocationDocument("example.com")
	revocation.AddRevokedKey(rev, "sha256:bad", revocation.ReasonKeyCompromise)
	data, _ := json.Marshal(rev)
	if err := os.WriteFile(filepath.Join(tmpDir, "example.com.revocations.json"), data, 0644); err != nil {
		t.Fatalf("failed to write file: %v", err)
	}

	resolver := NewLocalFileResolver(".", tmpDir)
	doc, err := resolver.ResolveRevocation("example.com", nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if doc == nil {
		t.Fatal("expected non-nil revocation")
	}
	if doc.Domain != "example.com" {
		t.Errorf("expected domain example.com, got %s", doc.Domain)
	}
}

func TestLocalFileResolverNoRevDir(t *testing.T) {
	resolver := NewLocalFileResolver(".", "")
	doc, err := resolver.ResolveRevocation("example.com", nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if doc != nil {
		t.Error("expected nil for empty revocation dir")
	}
}

func TestChainResolverFirstWins(t *testing.T) {
	b1 := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	b1.Documents = append(b1.Documents, bundle.BundledDiscovery{
		Domain: "a.com",
		WellKnown: discovery.WellKnownResponse{
			SchemaVersion: "1.2",
			DeveloperName: "First",
			PublicKeyPEM:  "PEM1",
		},
	})

	b2 := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	b2.Documents = append(b2.Documents, bundle.BundledDiscovery{
		Domain: "a.com",
		WellKnown: discovery.WellKnownResponse{
			SchemaVersion: "1.2",
			DeveloperName: "Second",
			PublicKeyPEM:  "PEM2",
		},
	})

	chain := NewChainResolver([]SchemaResolver{
		NewTrustBundleResolver(b1),
		NewTrustBundleResolver(b2),
	})

	disc, err := chain.ResolveDiscovery("a.com")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if disc.DeveloperName != "First" {
		t.Errorf("expected First, got %s", disc.DeveloperName)
	}
}

func TestChainResolverFallthrough(t *testing.T) {
	b1 := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	b1.Documents = append(b1.Documents, bundle.BundledDiscovery{
		Domain: "a.com",
		WellKnown: discovery.WellKnownResponse{
			SchemaVersion: "1.2",
			DeveloperName: "First",
			PublicKeyPEM:  "PEM1",
		},
	})

	b2 := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	b2.Documents = append(b2.Documents, bundle.BundledDiscovery{
		Domain: "b.com",
		WellKnown: discovery.WellKnownResponse{
			SchemaVersion: "1.2",
			DeveloperName: "Second",
			PublicKeyPEM:  "PEM2",
		},
	})

	chain := NewChainResolver([]SchemaResolver{
		NewTrustBundleResolver(b1),
		NewTrustBundleResolver(b2),
	})

	disc, err := chain.ResolveDiscovery("b.com")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if disc.DeveloperName != "Second" {
		t.Errorf("expected Second, got %s", disc.DeveloperName)
	}
}

func TestChainResolverAllMiss(t *testing.T) {
	b := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	chain := NewChainResolver([]SchemaResolver{
		NewTrustBundleResolver(b),
	})

	_, err := chain.ResolveDiscovery("missing.com")
	if err == nil {
		t.Error("expected error when all resolvers miss")
	}
}

File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/javascript/tests/resolver.test.js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { writeFileSync, mkdtempSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';

import {
    TrustBundleResolver,
    LocalFileResolver,
    ChainResolver
} from '../src/resolver.js';
import { createTrustBundle, createBundledDiscovery } from '../src/bundle.js';
import {
    buildRevocationDocument,
    addRevokedKey,
    RevocationReason
} from '../src/revocation.js';

function makeBundle() {
    const wellKnown = {
        schema_version: '1.2',
        developer_name: 'Test Dev',
        public_key_pem: '-----BEGIN PUBLIC KEY-----\ntest\n-----END PUBLIC KEY-----'
    };
    const doc = createBundledDiscovery('example.com', wellKnown);
    const rev = buildRevocationDocument('example.com');
    addRevokedKey(rev, 'sha256:old', RevocationReason.SUPERSEDED);

    const bundle = createTrustBundle('2026-01-01T00:00:00Z');
    bundle.documents.push(doc);
    bundle.revocations.push(rev);
    return bundle;
}

describe('TrustBundleResolver', () => {
    it('should resolve discovery for known domain', async () => {
        const resolver = new TrustBundleResolver(makeBundle());
        const disc = await resolver.resolveDiscovery('example.com');
        assert.ok(disc);
        assert.equal(disc.developer_name, 'Test Dev');
    });

    it('should return null for unknown domain', async () => {
        const resolver = new TrustBundleResolver(makeBundle());
        assert.equal(await resolver.resolveDiscovery('unknown.com'), null);
    });

    it('should resolve revocation for known domain', async () => {
        const resolver = new TrustBundleResolver(makeBundle());
        const disc = await resolver.resolveDiscovery('example.com');
        const rev = await resolver.resolveRevocation('example.com', disc);
        assert.ok(rev);
        assert.equal(rev.domain, 'example.com');
        assert.equal(rev.revoked_keys.length, 1);
    });

    it('should create from JSON', async () => {
        const bundle = makeBundle();
        const json = JSON.stringify(bundle);
        const resolver = TrustBundleResolver.fromJson(json);
        const disc = await resolver.resolveDiscovery('example.com');
        assert.ok(disc);
    });
});

describe('LocalFileResolver', () => {
    it('should resolve discovery from file', async () => {
        const tmpDir = mkdtempSync(join(tmpdir(), 'schemapin-test-'));
        const wellKnown = {
            schema_version: '1.2',
            developer_name: 'File Dev',
            public_key_pem: 'PEM_DATA'
        };
        writeFileSync(join(tmpDir, 'example.com.json'), JSON.stringify(wellKnown));

        const resolver = new LocalFileResolver(tmpDir);
        const disc = await resolver.resolveDiscovery('example.com');
        assert.ok(disc);
        assert.equal(disc.developer_name, 'File Dev');
    });

    it('should return null for missing file', async () => {
        const tmpDir = mkdtempSync(join(tmpdir(), 'schemapin-test-'));
        const resolver = new LocalFileResolver(tmpDir);
        assert.equal(await resolver.resolveDiscovery('missing.com'), null);
    });

    it('should resolve revocation from file', async () => {
        const tmpDir = mkdtempSync(join(tmpdir(), 'schemapin-test-'));
        const rev = buildRevocationDocument('example.com');
        addRevokedKey(rev, 'sha256:bad', RevocationReason.KEY_COMPROMISE);
        writeFileSync(join(tmpDir, 'example.com.revocations.json'), JSON.stringify(rev));

        const resolver = new LocalFileResolver('.', tmpDir);
        const revocation = await resolver.resolveRevocation('example.com', {});
        assert.ok(revocation);
        assert.equal(revocation.domain, 'example.com');
    });

    it('should return null when no revocation dir', async () => {
        const resolver = new LocalFileResolver('.');
        assert.equal(await resolver.resolveRevocation('example.com', {}), null);
    });
});

describe('ChainResolver', () => {
    it('should return first match', async () => {
        const bundle1 = createTrustBundle('2026-01-01T00:00:00Z');
        bundle1.documents.push(createBundledDiscovery('a.com', {
            schema_version: '1.2',
            developer_name: 'First',
            public_key_pem: 'PEM1'
        }));

        const bundle2 = createTrustBundle('2026-01-01T00:00:00Z');
        bundle2.documents.push(createBundledDiscovery('a.com', {
            schema_version: '1.2',
            developer_name: 'Second',
            public_key_pem: 'PEM2'
        }));

        const chain = new ChainResolver([
            new TrustBundleResolver(bundle1),
            new TrustBundleResolver(bundle2)
        ]);
        const disc = await chain.resolveDiscovery('a.com');
        assert.equal(disc.developer_name, 'First');
    });

    it('should fallthrough to second resolver', async () => {
        const bundle1 = createTrustBundle('2026-01-01T00:00:00Z');
        bundle1.documents.push(createBundledDiscovery('a.com', {
            schema_version: '1.2',
            developer_name: 'First',
            public_key_pem: 'PEM1'
        }));

        const bundle2 = createTrustBundle('2026-01-01T00:00:00Z');
        bundle2.documents.push(createBundledDiscovery('b.com', {
            schema_version: '1.2',
            developer_name: 'Second',
            public_key_pem: 'PEM2'
        }));

        const chain = new ChainResolver([
            new TrustBundleResolver(bundle1),
            new TrustBundleResolver(bundle2)
        ]);
        const disc = await chain.resolveDiscovery('b.com');
        assert.ok(disc);
        assert.equal(disc.developer_name, 'Second');
    });

    it('should return null if all miss', async () => {
        const bundle = createTrustBundle('2026-01-01T00:00:00Z');
        const chain = new ChainResolver([new TrustBundleResolver(bundle)]);
        assert.equal(await chain.resolveDiscovery('missing.com'), null);
    });
});
